### ISSUE TITLE
PermissionRequiredMixin should have option to redirect if anonymous but also raise PermissionDenied if authenticated

### ISSUE THREAD

**Issue by spuriousdata** [django-guardian/django-guardian#200]
I was going to fix this and submit a patch, but in looking through the code it seems like more of a missing feature than a bug. If you have return_403 and raise_exception set to False, the code will redirect you to the login page even if you're already logged in but simply don't have permission. If you set either return_403 or raise_exception, it does just the opposite which is (effectively) render the 403 template even if you're not logged in. Shouldn't there be an option (possibly a third one) that raises an exception or renders the 403 template while still redirecting you to the login page if you are not logged in?

**Comment by hanneshapke** [comment]
+1

**Comment by brianmay** [comment]
Probably no point redirecting to the login page if you are already logged in. Maybe we should force return_403 to True if the user is not logged in? Or the reverse: force it to False if the user is logged in. Not sure if either would break anything...

**Comment by brianmay** [comment]
So the code in question is:

```
    if not has_permissions:
        if return_403:
            if guardian_settings.RENDER_403:
                try:
                    response = render_to_response(
                        guardian_settings.TEMPLATE_403, {},
                        RequestContext(request))
                    response.status_code = 403
                    return response
                except TemplateDoesNotExist as e:
                    if settings.DEBUG:
                        raise e
            elif guardian_settings.RAISE_403:
                raise PermissionDenied
            return HttpResponseForbidden()
        else:
            return redirect_to_login(request.get_full_path(),
                                     login_url,
                                     redirect_field_name)
```

I am not sure why we need so many options. Seems to be overly complicated.

**Comment by brianmay** [comment]
So if return_403 is set, we will return an rendered HttpResponse or an PermissionDenied or an ttpResponseForbidden()

If return_403 is not set, we return a redirect_to_login instead.

**Comment by brianmay** [comment]
Am inclined to think that we shouldn't be capturing the TemplateDoesNotExist exception. I think this is an error regardless of whether DEBUG is enabled or not, and we shouldn't try to hide it.

**Comment by brianmay** [comment]
In check_permissions(), if return_403 is set but RENDER_403 and RAISE_403 are false, the get_403_or_None call will generate a PermissionDenied exception, resulting in on_permission_check_fail() not being called.

**Comment by felixrindt** [comment]
Thanks for mentioning us @thclark! 
We subclassed djangos stock mixin to get the functionality we need (not optimized for speed or anything):
https://github.com/ephios-dev/ephios/blob/c0712411bce3da183afc991369a6697a191d19fb/ephios/extra/mixins.py#L9-L36

You mentioned wanting a non-breaking PR for this issue, but judging from the other comments here, we can't just redefine the meaning of existing flags on the Mixin without breaking existing behavior. An extra flag seems unclean to me, too. 
I'm not too familiar with the internals of how the existing flags work and it's to far into the night to wrap my head around it 😅   

@brianmay Did you perhaps implement this with the guardian mixin?

**Comment by Natgho** [comment]
Hi @felixrindt , this ticket is coming from ancient times (=

We currently have a permission mixin within the existing shortcuts;
https://django-guardian.readthedocs.io/en/stable/api/mixins/#guardian.mixins.PermissionRequiredMixin

I believe you needed this at the time. Have I understood correctly?

If so, I think we can close this ticket.

**Comment by spuriousdata** [comment]
@felixrindt That is also what I did since I didn't have 11 years to wait for it to be fixed. Eventually is better than never, though.

**Comment by felixrindt** [comment]
> I believe you needed this at the time. Have I understood correctly?

I just checked again, [the initial issue here](https://github.com/django-guardian/django-guardian/issues/200#issue-29014206) is still not addressed. From what I can tell, the docs saying "If the user isn't logged in, redirect to settings.LOGIN_URL" is misleading, because independent of whether is logged in or not, the mixin will use `get_40x_or_None`, which determines its response solely based on its input arguments (`return_403`, `return_404`, ...) and the classes `raise_exception` attribute. 

To get the behavior this issues demands (redirecting anonymous users and having PermissionDenied raised for logged in users), I now figured this small modification to the Mixin does the trick:

```python
class CustomPermissionRequiredMixin(GuardianPermissionRequiredMixin):
    """
	Raises PermissionDenied for authenticated users, redirects anonymous users to login.
	"""

    @property
    def raise_exception(self):
        return self.request.user.is_authenticated
```

**Comment by Natgho** [comment]
> [@felixrindt](https://github.com/felixrindt) That is also what I did since I didn't have 11 years to wait for it to be fixed. Eventually is better than never, though.

Of course, since this is a project based on volunteerism, should we be surprised that no one has come forward in 11 years, or is it right to be angry about it? :) It's very difficult these days for anyone to put their hand in the fire... Everyone prefers to be in the consumer position, even as developers...

Whatever, Let's focus on our work.
> I just checked again, [the initial issue here](https://github.com/django-guardian/django-guardian/issues/200#issue-
> ...
>     @property
>     def raise_exception(self):
>         return self.request.user.is_authenticated

 @felixrindt This looks great and makes sense, but instead of overriding every exception, wouldn't it be better to just override the None check or the 404? If you submit a PR like this, we'd be happy to review and accept it.

**Comment by felixrindt** [comment]
> This looks great and makes sense, but instead of overriding every exception, wouldn't it be better to just override the None check or the 404? If you submit a PR like this, we'd be happy to review and accept it.

Great! It's not that I don't want to create a PR - I got the repo forked and set up. I just wasn't sure how to go about it, but I think now I have an action plan for a backwards compatible change: 😅 

Because anonymous users might also have permission, we must determine the "forbidden" condition first as part of the actual permission check in `get_40x_or_None` and only then return a redirect-to-login for anonymous users. 

If we put the check after the call to an unchanged `get_40x_or_None`, it wont work for the case that programmers use the `RAISE_40x` setting, which might be desirable for handling "forbidden" for authenticated users. I don't think fumbling with the `return_40x` or `raise_40x` options based on the user being authenticated is a nice solution, seems even more spaghetti to me.

So, a check for anonymous users would ideally have to be after the permission check but before any exceptions are raised, which is inside `get_40x_or_None`. Therefore a fix might best be placed inside that functions flow:

- We introduce a new kwarg `get_40x_or_None(..., redirect_anonymous=False)` keeping it backwards compatible. If this kwarg is set to True, for anonymous users, the `return_403` and `return_404` branches will be skipped and a redirect will be returned.
- In the `PermissionRequiredMixin`, we could then simply have a `redirect_anonymous = False` as an attribute, pass it to `get_40x_or_None` (just like `return_403` etc.)  and programmers could set it to True if they want that behavior (update the docs accordingly).
- Same simple change would work on the `permission_required` decorator.

It does introduce even more complexity to `get_40x_or_None`, but I think with some refactoring it might still be equally understandable. I can whip up the PR if the plan sounds good to you 🎉

**Comment by Natgho** [comment]
> Great! It's not that I don't want to create a PR - I got the repo forked and set up. I just wasn't sure how to go about it, but I think now I have an action plan for a backwards compatible change: 😅
> ...
> It does introduce even more complexity to `get_40x_or_None`, but I think with some refactoring it might still be equally understandable. I can whip up the PR if the plan sounds good to you 🎉

Haha, can we call you the crazy software developer who forked it and created his own version because he didn't know how to create PR? :)

This idea structurally provide more options and provides a more detailed control structure for 40x errors. Of course make sense for me, please document it thoroughly and explain the “how” nicely so that developers don't get confused when using it.

Also, please consider **backward compatibility**(The default behavior should remain unchanged) so that projects upgrading to the new version with the update aren't affected.

I'll be eagerly awaiting your PR 🚀

**Comment by spuriousdata** [comment]
@Natgho I'm not mad, but if anyone had bothered even replying to confirm that a fix was desired in less than two full years from when I submitted the ticket, then I would have fixed it and submitted a pull request. At this point, the project this was for is long gone and the company it was for no longer exists.

**Comment by Natgho** [comment]
> [@Natgho](https://github.com/Natgho) I'm not mad, but if anyone had bothered even replying to confirm that a fix was desired in less than two full years from when I submitted the ticket, then I would have fixed it and submitted a pull request. At this point, the project this was for is long gone and the company it was for no longer exists.

I understand. That's very sad. But we can think of it this way: at least now someone is trying to take care of django-guardian like it's their baby :)

If you have time, we'd be happy to have you join us.
